Improve performance, streamline code in a LOT of places. Fixed bugs.

laundmo opened the pull request:
Bugs fixed:
- can only have a single reloading loop - now works with multiple
- loop vars if reloading loop isn't first were wrong. - dumb regex looked at the first loop it could find, now getting vars using ast.

Performance improvements:
- now allows for the keyword only argument reload_after, which will only reload the code every n iterations
- reduced total number of inspect calls
- compile() the code first, before running it

Reloading loop performance comparison: 
![image](https://user-images.githubusercontent.com/24855949/109418254-b65e0580-79c7-11eb-87e5-272b8fbff841.png)
`r` is the reload_after
`old` refers to whether this is the old version of the library.

The benchmark code is avaliable on [laundmo:benchmark](https://github.com/laundmo/reloading/tree/benchmark)

This is heavily related to #10

laundmo committed:
Improve performance, streamline code in a LOT of places. Fixed bugs.

Bugs fixed:
- can only have a single reloading loop - now works with multiple
- loop vars if reloading loop isn't first were wrong. - dumb regex looked at the first loop it could find, now getting vars using ast.

Performance improvements:
- now allows for the keyword only argument reload_after, which will only reload the code every n iterations
- reduced total number of inspect calls
- compile() the code first, before running it (only loops so far)

laundmo committed:
clean up reloading function logic, added test case

julvo commented:
Hey, thanks a lot for the PR 🙏 Will have a look through the code sometime this week

laundmo commented:
hey, i need to test this some more, i remembered that i haven't tested how well errors are handled, more specifically, syntax errors.
i moved the compilation step outside of the loop/decorator, which may mean that syntax errors happen there and aren't caught by the existing error handler.

I will test this over the next few days, and comment the results.

sidenote: i use the black formatter, which is why the diff probably seems larger than expected. didn't mean to run it, but oh well.

laundmo committed:
rework function code to be ast only, no more line numbers

laundmo commented:
@julvo i have now reworked the way functions reload (no more loading specific line numbers!) and also fixed the issues with error handling, which was also streamlined some more. Its definitely ready for merge from my side, please do still test/review it.

at this point i've basically rewritten the entire thing 😆 

EDIT: scratch that, i broke the normal looping. back to work.

EDIT: EDIT: fixed it

laundmo committed:
fix dumb mistake breaking normal loops

laundmo committed:
vastly simplify and improve loop code.
no more string splitting and loading the source multiple files.

laundmo committed:
Add easy infinite loops by pasing True or a int.

laundmo committed:
Explain new functionality in README

laundmo commented:
I think I'm finally done. Unless I think of some amazing way to make it work with while loops, but there most likely isn't one.

laundmo commented:
@julvo whats the status on this?

julvo commented:
I'm currently quite busy, but hope to get around to it it on the weekend. As it's quite a substantial change, I want to review and test this properly. Generally open to extending the API, but may want to change the naming a bit. Initial idea for the new API is `reloading(every=10, forever=True)` instead of `reloading(1, reload_after=10)` and removing the option to do `reloading(True)`

laundmo commented:
@julvo thats fine, and definitely not too hard to implement, will take a look today or tomorrow on changing that. i should probably also write some more unit tests for that new behavior.

Generally, the infinite looping definitely isn't required and if that's what's holding up this PR from being merged, i would rather remove it for now and do a second PR with those features.

laundmo committed:
update with suggested API changes
also add handling of empty reloading in loop

laundmo commented:
Finally got around to implementing the suggested changes.
I also went ahead and added handling for people using a empty call to reloading() in their loop - will now raise a proper error instead of partial something.

laundmo committed:
Represent API changes in readme

julvo commented:
Hey @laundmo, had a look at some of the patch last weekend. So far only the reloading decorator code - good job! I like the idea of filtering out the reloading decorator instead of remembering the inner function explicitly. Just wanted to let you know that I'm not forgetting to review this, just quite busy at the moment

laundmo commented:
thanks for the notification, i was wondering about that.
take your time!

laundmo commented:
more than a month later, i already have quite a few people installing reloading directly from my fork, so i might just re-release under a new name because this is taking entirely too long. Would also allow me to do some more drastic changes.

laundmo committed:
changed install instructions

julvo commented:
Just merged this. Sorry again for taking so long - it has been a busy period for me. Of course, feel free to maintain your own fork or also open issues on this repo

closes #10

laundmo commented:
thanks, i created some issues to keep track of things that need doing, mostly so i wont forget
